src/main/scala/chisel3/util/SRAM.scala

```diff
@@ -470,54 +472,140 @@ object SRAM {
       )
     }
 
-    val mem = SyncReadMem(size, tpe)
-    val _out = Wire(new SRAMInterface(size, tpe, numReadPorts, numWritePorts, numReadwritePorts, isVecMem))
+    // underlying target
+    val mem = new SramTarget()
+
+    // user-facing interface into the SRAM
+    val sramIntfType = new SRAMInterface(size, tpe, numReadPorts, numWritePorts, numReadwritePorts, isVecMem)
+    val _out = Wire(sramIntfType)
     _out._underlying = Some(HasTarget(mem))
 
-    for ((clock, port) <- readPortClocks.zip(_out.readPorts)) {
-      port.data := mem.read(port.address, port.enable, clock)
+    // create actual ports into firrtl memory
+    val firrtlReadPorts:  Seq[FirrtlMemoryReader[_]] = sramIntfType.readPorts.map(new FirrtlMemoryReader(_))
+    val firrtlWritePorts: Seq[FirrtlMemoryWriter[_]] = sramIntfType.writePorts.map(new FirrtlMemoryWriter(_))
+    val firrtlReadwritePorts: Seq[FirrtlMemoryReadwriter[_]] =
+      sramIntfType.readwritePorts.map(new FirrtlMemoryReadwriter(_))
+
+    // set references to firrtl memory ports
+    val firrtlReadPortNames: Seq[String] = firrtlReadPorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"R$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
+    val firrtlWritePortNames: Seq[String] = firrtlWritePorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"W$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
+    val firrtlReadwritePortNames: Seq[String] = firrtlReadwritePorts.zipWithIndex.map {
+      case (p, idx) =>
+        val name = s"RW$idx"
+        p.setRef(Slot(Node(mem), name))
+        name
+    }
```

Comment:
```suggestion
    def nameAndSetRef(ports: Seq[Data], namePrefix: String): Seq[Name] = {
      ports.zipWithIndex.map { case (p, idx) =>
        val name = namePrefix + idx
        p.setRef(Slot(Node(mem), name))
        name        
      }
    }
    val firrtlReadPortNames: Seq[String] = nameAndSetRef(firrtlReadPorts, "R")
    val firrtlWritePortNames: Seq[String] = nameAndSetRef(firrtlWritePorts, "W")
    val firrtlReadwritePortNames: Seq[String] = nameAndSetRef(firrtlReadwritePorts, "RW")
```
Reduce, reuse, recycle